compiler/optimizations/preNormalizeOptimizations.cpp

```diff
@@ -102,36 +117,256 @@ void doPreNormalizeArrayOptimizations() {
-        return confirmAccess(call);
-      }
-      else {
-        return revertAccess(call);
+    ret = confirmed ? confirmAccess(call) : revertAccess(call);
+  }
+  else {
+    // maybe we have automatic aggregation but no automatic local access? In
+    // that case we may have generated one of this primitive, for aggregation
+    // detection, now we just remove it silently
+    ret = revertAccess(call);
+  }
+
+  INT_ASSERT(ret);
+
+  if (fAutoAggregation) {
+    AggregationCandidateInfo *aggCandidate = preNormalizeAggCandidate[call];
+    if (aggCandidate != NULL) {
+
+      if (fReportAutoAggregation) {
+        std::stringstream message;
+        message << "Aggregation candidate ";
+        message << "has ";
+        message << (confirmed ? "confirmed" : "reverted");
+        message << " local child ";
+        message << getForallCloneTypeStr(aggCandidate->forall);
+        LOG_AA(0, message.str().c_str(), aggCandidate->candidate);
       }
+
+      aggCandidate->logicalChildAnalyzed(call, confirmed);
     }
-    else {
-      INT_FATAL("Misconfigured PRIM_MAYBE_LOCAL_THIS");
+  }
+  return ret;
+}
+
+// called during LICM to restructure a conditional aggregation to direct
+// aggregation
+void transformConditionalAggregation(CondStmt *cond) {
+  // move the aggregation call before the conditional (at this point in
+  // compilation it must be inlined)
+  for_alist(expr, cond->elseStmt->body) {
+    cond->insertBefore(expr->remove());
+  }
+  
+  // remove the defpoint of the aggregation marker
+  SymExpr *condExpr = toSymExpr(cond->condExpr);
+  INT_ASSERT(condExpr);
+
+  if (fReportAutoAggregation) {
+    std::stringstream message;
+    message << "Replaced assignment with aggregation";
+    if (condExpr->symbol()->hasFlag(FLAG_AGG_IN_STATIC_AND_DYNAMIC_CLONE)) {
+      message << " [static and dynamic ALA clone] ";
     }
+    else if (condExpr->symbol()->hasFlag(FLAG_AGG_IN_STATIC_ONLY_CLONE)) {
+      message << " [static only ALA clone] ";
+    }
+    LOG_AA(0, message.str().c_str(), condExpr);
   }
-  return NULL;
+
+  condExpr->symbol()->defPoint->remove();
+
+  // remove the conditional
+  cond->remove();
 }
 
+// called after unordered forall optimization to remove aggregation code (by
+// choosing non-aggregated version)
+void cleanupRemainingAggCondStmts() {
+  forv_Vec(CondStmt, condStmt, gCondStmts) {
+    if (condStmt->inTree()) {
+      if (SymExpr *condSymExpr = toSymExpr(condStmt->condExpr)) {
+        //std::cout << "100\n";
+        if (condSymExpr->symbol()->hasFlag(FLAG_AGG_MARKER)) {
+          // this is a conditional that wasn't modified by the unordered
+          // optimization, so we need to remove it and the aggregator associated
+          // with it
+
+          // find the aggregator symbol within the else block
+          // note that the else block will be inlined and will be a big mess, we
+          // need to find the compiler-generated aggregator in that mess
```

Comment:
The comments don't actually say it - but I'm understanding/guessing that this is about removing the effects of the forall intent adding the aggregator? Could you make the comments clearer about that? Seems like we could set its type to `dtNothing` and expect other dead code elimination to handle it, in addition/alternatively.

Reply:
I'll update the comments.

For `dtNothing`: interesting.. before I wrote this part, I was hoping that I can just leave the code around, and as the aggregator symbol is not used, it'll be removed as dead code. But that didn't happen. My thinking was this was happening too late in compilation and there was no pruning happening. But if I were to set the type to `dtNothing`, it would have been cleaned?

Maybe it didn't happen, because def/use analysis was challenging at that point, or maybe the destruction of the symbol at the end was seen as a "use".